test/end2end_test.go

```diff
@@ -7032,6 +7032,84 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 	}()
 }
 
+// TestClientTransportRestartsAfterStreamIdExhausted tests that the client transport
+// drains and restarts when next stream ID exceeds MaxStreamID. This test also
+// verifies that subsequent RPCs use a new client transport and the old transport
+// is closed.
+func (s) TestClientTransportRestartsAfterStreamIdExhausted(t *testing.T) {
+	// overriding MaxStreamIDForTesting.
+	originalMaxStreamID := transport.MaxStreamIDForTesting
+	transport.MaxStreamIDForTesting = 5
+	defer func() {
+		transport.MaxStreamIDForTesting = originalMaxStreamID
+	}()
+
+	// setting up StubServer.
+	s := grpc.NewServer()
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for i := 0; i < 10; i++ {
+				m := &grpc.PreparedMsg{}
+				stream.SendMsg(m)
+			}
+			return nil
+		},
+	}
+	testpb.RegisterTestServiceServer(s, ss)
+
+	// setting up gRPC server with ListenerWrapper.
+	lisWrap := testutils.NewListenerWrapper(t, nil)
+	go s.Serve(lisWrap)
+	defer s.Stop()
+
+	cc, err := grpc.Dial(lisWrap.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("dial(%q): %v", lisWrap.Addr().String(), err)
+	}
+	defer cc.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+	defer cancel()
+
+	client := testpb.NewTestServiceClient(cc)
+
+	createStreamAndRecv := func() {
+		stream, err := client.FullDuplexCall(ctx)
+		if err != nil {
+			t.Fatalf("creating FullDuplex stream: %v", err)
+		}
+		if _, err = stream.Recv(); err != nil && err != io.EOF {
+			t.Fatalf("stream.Recv() = _, %v want: nil or EOF", err)
+		}
+	}
+
+	// creating FullDuplexCall stream #1.
+	createStreamAndRecv()
+	// creating FullDuplexCall stream #2.
+	createStreamAndRecv()
+
+	// verifying creation of new conn channel.
+	val, err := lisWrap.NewConnCh.Receive(ctx)
+	if err != nil {
+		t.Fatal("timeout expired when waiting to create new conn channel")
+	}
+	conn1 := val.(*testutils.ConnWrapper)
+
+	// this stream should be created in a new conn channel.
+	createStreamAndRecv()
+
+	// verifying a new conn channel is created.
+	if _, err = lisWrap.NewConnCh.Receive(ctx); err != nil {
+		t.Fatal("timeout expired when waiting to create new conn channel")
+	}
+
+	// verifying the connection to the old one is drained and closed.
```

Comment:
Interestingly you haven't fully received off the streams that were active.  So this only works because the flow control is enough that the server finished the stream.

I'd recommend avoiding "coincidences" like this (even if they're literally never going to not be true).

Instead you should make the client & server behavior align.

It also could be interesting to test that the stream is still active even after the new connection is created, which is how "draining" behavior should work.  E.g.

1. Create a stream that ping-pongs back and forth.
2. Use it and make sure it works.
3. Create a stream on the next connection that does the same.  (Verify that the new connection was used for it somehow.)
4. Use it and make sure it works.
5. Verify that the stream from (1) still works.
6. Close streams from (1) and (3).  Client-side context cancelation should be easy and fine here.
7. Expect the original connection to close.

One way to validate which connection the stream is using: you could use a custom credentials (its AuthInfo ends up in the Peer) that counts connections and increments some number for each one.  Creds could also be used to determine when connections are closed, since they return a wrapped `net.Conn` back to us that is what we really close.

Reply:
thanks for the detailed comment. I've updated the test to align to your comment. lemme know what you think.